code/modules/antagonists/traitor/objectives/sabotage_machinery.dm

```diff
@@ -0,0 +1,133 @@
+/datum/traitor_objective_category/sabotage_machinery
+	name = "Sabotage Worksite"
+	objectives = list(
+		/datum/traitor_objective/sabotage_machinery = 1,
+	)
+
+/datum/traitor_objective/sabotage_machinery
+	name = "Sabotage the %MACHINE%"
+	description = "Conceal the provided explosive device within the %MACHINE% to cause disarray and disrupt the operations of the %JOB%'s department."
+
+	progression_reward = list(2 MINUTES, 8 MINUTES)
+	telecrystal_reward = list(0, 1)
+
+	progression_minimum = 0 MINUTES
+	progression_maximum = 10 MINUTES
+
+	/// The maximum amount of this type of objective a traitor can have.
+	var/maximum_allowed = 2
+	/// The possible target machinery and the jobs tied to each one.
+	var/list/applicable_jobs = list(
+		JOB_CHIEF_ENGINEER = /obj/machinery/rnd/production/protolathe/department/engineering,
+		JOB_CHIEF_MEDICAL_OFFICER = /obj/machinery/rnd/production/techfab/department/medical,
+		JOB_HEAD_OF_PERSONNEL = /obj/machinery/rnd/production/techfab/department/service,
+		JOB_QUARTERMASTER = /obj/machinery/rnd/production/techfab/department/cargo,
+		JOB_RESEARCH_DIRECTOR = /obj/machinery/rnd/production/protolathe/department/science,
+		JOB_SHAFT_MINER = /obj/machinery/mineral/ore_redemption,
+	)
+	/// The chosen job. Used to check for duplicates
+	var/chosen_job
+
+	var/obj/item/traitor_machine_trapper/tool
+
+/datum/traitor_objective/sabotage_machinery/can_generate_objective(datum/mind/generating_for, list/possible_duplicates)
+	if(length(possible_duplicates) >= maximum_allowed)
+		return FALSE
+	return TRUE
+
+/datum/traitor_objective/sabotage_machinery/generate_objective(datum/mind/generating_for, list/possible_duplicates)
+	var/list/possible_jobs = applicable_jobs.Copy()
+	for(var/datum/traitor_objective/sabotage_machinery/objective as anything in possible_duplicates)
+		possible_jobs -= objective.chosen_job
+	if(!length(possible_jobs))
+		return FALSE
+	var/list/obj/machinery/possible_machines = list()
+	while(length(possible_machines) <= 0 && length(possible_jobs) > 0)
+		var/target_head = pick(possible_jobs)
+		var/obj/machinery/machine_to_find = possible_jobs[target_head]
+		possible_jobs -= target_head
+
+		chosen_job = target_head
+		for(var/obj/machinery/machine as anything in GLOB.machines)
+			if(istype(machine, machine_to_find) && is_station_level(machine.z))
+				possible_machines += machine
```

Comment:
You shouldn't be generating the objectives like this, this is laggy.
Generate a list of possible targets like how the steal objective works and access the types from this list.

Reply:
nvm you can ignore this

Reply:
Too late I agree so I have written it